### PR TITLE
Add feature to ingest vector datasets

### PR DESCRIPTION
**Pull request by smohiudd**
Includes:

0. new `submit-vector` lambda and associated step function. Submit vector also include security group, RDS VPC lookup and database ingress.
0. `psf/black` pre-commit setup to avoid manual labor
0. `.toml` and `poetry` changes to fix CI failure for tests around `smart-open` and `pypgstac`

One of our tests is failing b/c `poetry` can't install `smart-open`. Doesn't affect merging of this MR or functionality.

### CODE REVIEW

File: deploy/app.py [hunk 1]
```diff
@@ -49,5 +50,9 @@
     lambda_function=lambda_stack.trigger_ingest_lambda,
     workflow_arn=pub_arn,
 )
+lambda_stack.grant_execution_privileges(
+    lambda_function=lambda_stack.trigger_vector_lambda,
+    workflow_arn=pub_arn,
```
Comment (alukach) [inline deploy/app.py:55]: I think this should be `vector_arn`

Comment (alukach) [reply deploy/app.py:55]:
```suggestion
    workflow_arn=vector_arn,
```

**Comment by ranchodeluxe** [comment]
@slesaad: I think this is ready for review. After some minor fixes last Thursday I was able to get the `submit-vector` step function flow to completely get to the point where `ogr` was trying to add it to the database. But I don't have the VPC endpoints set up for the test DB. I believe that means this puppy is ready for primetime. Thoughts?

Also, I'm used to `black` formatting code for me but I can't tell if the failing jobs above are just intermittent fails. Let me rerun.

**Comment by ranchodeluxe** [comment]
> @slesaad: I think this is ready for review. After some minor fixes last Thursday I was able to get the `submit-vector` step function flow to completely get to the point where `ogr` was trying to add it to the database. But I don't have the VPC endpoints set up for the test DB. I believe that means this puppy is ready for prime time. Thoughts?

File: .github/workflows/cicd.yml [hunk 2]
```diff
@@ -21,6 +21,10 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - uses: psf/black@stable
+        with:
+          options: "--check --verbose"
+          src: "./deploy ./scripts ./lambdas"
+          version: "22.8.0"
```
Comment (ranchodeluxe) [inline .github/workflows/cicd.yml:27]: reviewer context: making sure our CI runner is using same version of `psf/black` that our `poetry.lock` specifies and now that we also use locally in pre-commit hook

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+repos:
```
Comment (ranchodeluxe) [inline .pre-commit-config.yaml:1]: reviewer comment: adding black to `pre-commit` so it auto formats for us to remove manual labor

File: .github/workflows/test_python_lambda.yml [hunk 4]
```diff
@@ -12,20 +12,37 @@ jobs:
...
           python-version: "3.8" # NOTE: Should match Lambda runtime specified in CDK code
-          cache: "poetry"
-          cache-dependency-path: |
-            poetry.lock
-            ${{ inputs.path_to_lambda }}/requirements**.txt
 
-      - name: Install requirements
+      - name: Install poetry
+        run: |
+          python -m pip install poetry==1.3.2
```
Comment (sbquinlan) [inline .github/workflows/test_python_lambda.yml:21]: Would bumping the version in pyproject accomplish this?

File: lambdas/submit-vector/handler.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        secrets (dict): decrypted secrets in dict
+    """
+
+    # Create a Secrets Manager client
+    session = boto3.session.Session(region_name="us-west-2")
+    client = session.client(service_name="secretsmanager")
+
+    # In this sample we only handle the specific exceptions for the 'GetSecretValue' API.
+    # See https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html
+    # We rethrow the exception by default.
```
Comment (sbquinlan) [inline lambdas/submit-vector/handler.py:50]: delete?

File: deploy/cdk/step_function_stack.py [hunk 6]
```diff
@@ -148,6 +166,35 @@ def _cogify_workflow(
...
+        vector_task = self._lambda_task(
+            "Ingest Vector",
+            lambda_stack.vector_lambda,
+        )
+
+        # enqueue_task = self._sqs_task(
+        #     "Send cogified to stac-ready queue",
+        #     queue=queue_stack.stac_ready_queue,
+        #     input_path="$.Payload",
+        # )
```
Comment (sbquinlan) [inline deploy/cdk/step_function_stack.py:182]: delete?

**Review by sbquinlan — COMMENTED** [review]
Keep in mind I have no idea what I'm doing still. Added really nitty comments.

Sidenote: should we just remove the stac-ready queue if this step writes directly to the DB? That kind of breaks the current architecture a bit huh?

**Comment by sbquinlan** [comment]
Alright, rebased and reset this back to where it was before the build broke.

**Review by ranchodeluxe — APPROVED** [review]
woot, woot, let's do this

**Comment by ranchodeluxe** [comment]
@smohiudd : you have a merge conflict and you also need to rebase on `main`

File: deploy/cdk/queue_stack.py [hunk 7]
```diff
@@ -56,6 +56,24 @@ def __init__(
...
+            dead_letter_queue=sqs.DeadLetterQueue(
+                max_receive_count=5,
+                queue=self._queue(f"{construct_id}-vector-dlq"),
+            ),
+        )
+
+        lambda_stack.trigger_vector_lambda.add_event_source(
+            lambda_event_sources.SqsEventSource(
+                self.vector_queue,
+                batch_size=1,
```
Comment (slesaad) [inline deploy/cdk/queue_stack.py:71]: might want to increase the `batch_size`?

Comment (smohiudd) [reply deploy/cdk/queue_stack.py:71]: Good catch. Will increase to match the other queues.

File: lambdas/submit-vector/handler.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    )
+
+
+def handler(event, context):
+    href = event["s3_filename"]
+    collection = event["collection"]
+
+    downloaded_filepath = download_file(href)
+
+    load_to_featuresdb(downloaded_filepath, collection)
```
Comment (slesaad) [inline lambdas/submit-vector/handler.py:93]: is there a  way to check if this succeeded or failed?

**Review by slesaad — APPROVED** [review]
some comments, but otherwise looks good to me

todo: create a ticket for ingest using a python library instead of using subprocess call

File: lambdas/submit-vector/handler.py (hunk 8 above)
Comment (smohiudd) [reply lambdas/submit-vector/handler.py:93]: Will that check in the next revision.

**Pull request merged**